![image](competi.png)

#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

In [1]:
import numpy as np
import pandas as pd

import urllib.request
from PIL import Image

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [2]:
from sklearn.experimental import enable_hist_gradient_boosting
from sklearn.ensemble import HistGradientBoostingRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.preprocessing import OrdinalEncoder

In [3]:
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_poisson_deviance
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestRegressor

In [4]:


from sklearn.ensemble import VotingRegressor

In [6]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from catboost import Pool, CatBoostRegressor
from sklearn.svm import SVC
from sklearn.feature_selection import SelectKBest
from sklearn import svm, datasets
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.linear_model import HuberRegressor
from sklearn.metrics import mean_squared_error

In [7]:
from sklearn.kernel_ridge import KernelRidge

In [8]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
df = pd.read_csv("diamonds_train.csv", index_col=0)

In [9]:
def score_estimator(estimator, df_test):
    """Score an estimator on the test set."""
    y_pred = estimator.predict(df_test)

    print("MSE: %.3f" %
          mean_squared_error(df_test["price"], y_pred,
                             sample_weight=None))
    print("MAE: %.3f" %
          mean_absolute_error(df_test["price"], y_pred,
                              sample_weight=None))

    mask = y_pred > 0
    if (~mask).any():
        n_masked, n_samples = (~mask).sum(), mask.shape[0]
        print(f"WARNING: Estimator yields invalid, non-positive predictions "
              f" for {n_masked} samples out of {n_samples}. These predictions "
              f"are ignored when computing the Poisson deviance.")

    print("mean Poisson deviance: %.3f" %
          mean_poisson_deviance(df_test["price"][mask],
                                y_pred[mask],
                                sample_weight=None))


In [ ]:
#df = pd.get_dummies(df)
#df

In [ ]:
#lb_make = LabelEncoder()
#df['cut'] = lb_make.fit_transform(df['cut'])
#df['color'] = lb_make.fit_transform(df['color'])
#df['clarity'] = lb_make.fit_transform(df['clarity'])

In [ ]:
#onehotencoder = OneHotEncoder(sparse=False, handle_unknown="error", drop="first")
#ohe = pd.DataFrame(onehotencoder.fit_transform(df[["color"]]))
#df = df.join(ohe)
#df.rename(columns={0: "a", 1: "b", 2: "c", 3: "d", 4: "e", 5: "f"}, inplace=True)
#oht = pd.DataFrame(onehotencoder.fit_transform(df[["clarity"]]))
#df = df.join(oht)
#df.rename(columns={0: "g", 1: "h", 2: "i", 3: "j", 4: "k", 5: "m", 6: "o", 7: "o", 8: "p"}, inplace=True)
#df["carat"] = df["carat"]*100
#df["price"] = df["price"]/100
#df["x"] = df["x"]*9
#df["y"] = df["y"]*9
#df["z"] = df["z"]*9
df["x"] = df["x"] + df["y"] + df["z"]

In [ ]:
#X = np.array(df[["carat", "cut",  "color", "clarity", "depth",  "table", "x", "y", "z"]]) #without cut, depth
#y = np.array(df["price"])
#X.shape

In [12]:
df.head()

carat        cut color clarity  depth  table     x     y     z  price
0   1.21      Ideal     H     VS2   63.0   57.0  6.73  6.70  4.23   6134
1   0.28  Very Good     D    VVS2   64.0   56.0  4.14  4.17  2.66    532
2   0.42    Premium     F     VS1   61.2   58.0  4.86  4.82  2.96   1103
3   0.26      Ideal     H      IF   61.1   57.0  4.16  4.12  2.53    600
4   1.10       Good     G     SI1   63.4   57.0  6.52  6.55  4.14   4997

In [112]:
#df['carat'] = (df['carat'] - df['carat'].mean()) / df['carat'].std()
#df['x'] = (df['x'] - df['x'].mean()) / df['x'].std()
#df['y'] = (df['y'] - df['y'].mean()) / df['y'].std()
#df['z'] = (df['z'] - df['z'].mean()) / df['z'].std()

In [11]:
#df['carat'] = df['carat']*10
#X = df[["carat", "x", "y", "z"]]
#y = df[["price"]]
#train_data = X_train
#train_label = y_train
#test_dataX = X_test
#test_datay = y_test


#train_pool = Pool(train_data, 
                  #train_label,
                  #)
#test_pool = Pool(test_dataX,
                 #test_datay)

In [48]:
#X = df[["carat", "x", "y", "z"]]
#y = df[["price"]]

In [ ]:
#train_data = X_train
#train_label = y_train
#test_dataX = X_test
#test_datay = y_test


#train_pool = Pool(train_data, 
                  train_label,
                  )
#test_pool = Pool(test_dataX,
                 test_datay)

In [ ]:
#model2 = CatBoostRegressor(learning_rate=0.05, depth=6, l2_leaf_reg=1, iterations=150)

#grid = {'learning_rate': [0.05],
        'depth': [6],
        'l2_leaf_reg': [1]}

#grid_search_result = model2.grid_search(grid, 
                                       #X, 
                                       #y, 
                                      # plot=True) #cat_features=[1, 2, 3]

In [10]:
train, test = train_test_split(df, test_size=0.2, random_state=0)

In [17]:
tree_preprocessor = ColumnTransformer(
    [
        ("categorical", OrdinalEncoder(),
            ["cut", "color", "clarity"]),
        ("numeric", "passthrough",
            ["carat", "x", "y", "z"]),
    ],
    remainder="drop",
)
poisson_gbrt = Pipeline([
    ("preprocessor", tree_preprocessor),
    ("regressor", HistGradientBoostingRegressor(loss="poisson",                           #Aqui se puede poner diferentes regressors
                                                max_leaf_nodes=128)),
])
poisson_gbrt.fit(df_train, df_train["price"],
                 regressor__sample_weight=None)

print("Poisson Gradient Boosted Trees evaluation:")
score_estimator(poisson_gbrt, df)

Poisson Gradient Boosted Trees evaluation:
MSE: 102529.445
MAE: 143.551
mean Poisson deviance: 12.578


In [ ]:
#reg1 = GradientBoostingRegressor(random_state=1)
#reg2 = RandomForestRegressor(random_state=1)
#reg3 = LinearRegression()
#ereg = VotingRegressor(estimators=[('gb', reg1), ('rf', reg2), ('lr', reg3)])

In [18]:
h = poisson_gbrt.predict(df)

In [19]:
mean_squared_error(df['price'], h)

102529.44520299621

In [20]:
h = poisson_gbrt.predict(df)

In [21]:
h

array([ 6210.89666667,   558.41      ,  1105.66      , ...,
        1299.76      , 10302.06      ,  1187.627     ])

In [22]:
df2 = df.drop(columns=['price'])

In [23]:
df2["price"] = 0

In [24]:
df2

carat        cut color clarity  depth  table     x     y     z  price
0       1.21      Ideal     H     VS2   63.0   57.0  6.73  6.70  4.23      0
1       0.28  Very Good     D    VVS2   64.0   56.0  4.14  4.17  2.66      0
2       0.42    Premium     F     VS1   61.2   58.0  4.86  4.82  2.96      0
3       0.26      Ideal     H      IF   61.1   57.0  4.16  4.12  2.53      0
4       1.10       Good     G     SI1   63.4   57.0  6.52  6.55  4.14      0
...      ...        ...   ...     ...    ...    ...   ...   ...   ...    ...
40340   1.55    Premium     H     VS2   61.3   61.0  7.46  7.39  4.55      0
40341   0.36      Ideal     D     SI1   60.6   56.0  4.58  4.63  2.79      0
40342   0.57  Very Good     I     VS2   62.2   55.0  5.33  5.34  3.32      0
40343   1.01  Very Good     F      IF   59.6   62.0  6.47  6.56  3.88      0
40344   0.54      Ideal     E     SI2   60.4   57.0  5.33  5.27  3.20      0

[40345 rows x 10 columns]

In [25]:
h = poisson_gbrt.predict(df2)

In [217]:
h.shape

(40345,)

In [24]:
h

array([ 6259.42 ,   571.3  ,  1099.4  , ...,  1300.94 , 10281.635,
        1225.27 ])

**Importante:**

   - Si quitas columnas o creas columnas nuevas a partir de otras, o cualquier modificación column-wise tendrás que aplicarlo al dataset de `titanic_test.csv` de cara a hacer la predicción.
   - Si por lo contrario, decides por ejemplo, quitar los outliers o hacer un `dropna()`, o cualquier modificación row-wise eso NO PODRÁS (ni debes) aplicarlo al dataset de `titanic_test.csv` de cara a hacer la predicción. ¿Por qué? Porque si el conjunto de test tiene 50 observaciones (filas) la predicción se espera que tenga 50 filas.

#### ¿Y si lo que hago es estandarizar los datos o hacer un encoding, también se lo tengo que hacer al test antes de predecir?

In [5]:
# ¿qué opináis? 

-----------------------------------------------------------------------------------------------------------------

### Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

#### ¿Y si después del análisis exploratorio de mis datos llego a la conclusión de que puedo predecir el precio solo con las columnas `x`, `y` y `z`, también aplica al test?

In [8]:
# ¿qué opináis?

### 2. Dividir X_train, X_test, y_train, y_test

### 3. Asignar el modelo (vacío) a una variable
Aquí meteríais los parámetros. 

**Consejo**: Usa GridSearch y vuélvete loca o loco probando modelos e hiperparámetros.

Siempre hay tiempo para una historia:
https://catboost.ai/docs/concepts/python-reference_catboostregressor.html

### 4. Entrenar el modelo con X_train, y_train

### 5. Predecir con el modelo ya entrenado con X_test

In [27]:
predictions = poisson_gbrt.predict(df)
print(predictions)

[ 6210.89666667   558.41        1105.66       ...  1299.76
 10302.06        1187.627     ]


### 6. Sacar métricas, valorar el modelo 

Recuerdo que en la competición se va a evaluar con la métrica de RMSE.

-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

_Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `diamonds_test.csv` nos arriesgamos a que digamos, no salga lo esperado._

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `diamonds_train.csv` al completo.

**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**

Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `diamonds_test.csv` para predecir.

**¿De dónde saco `diamonds_test.csv`?**

In [29]:
X_pred = pd.read_csv("diamonds_test.csv", index_col=0)
X_pred.head()

carat        cut color clarity  depth  table     x     y     z
0   0.30      Ideal     H     SI2   60.0   56.0  4.41  4.43  2.65
1   0.34      Ideal     D      IF   62.1   57.0  4.52  4.46  2.79
2   1.57  Very Good     I     VS2   60.3   58.0  7.58  7.55  4.56
3   0.31      Ideal     H     VS2   61.8   57.0  4.32  4.36  2.68
4   1.51       Good     I    VVS1   64.0   60.0  7.26  7.21  4.63

In [30]:
X_pred["price"] = 0

In [31]:
X_pred.shape

(13449, 10)

**¿Por qué da error?** 

In [32]:
predictions_submit = poisson_gbrt.predict(X_pred)
predictions_submit

array([  419.55,  1777.  ,  9852.21, ..., 15438.9 , 10677.24,   720.99])

## IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 13449 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `diamonds_test.csv`?**

In [ ]:
# ¿Qué opináis?

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

**¿Qué más habrá que quitar o hacer?**

In [ ]:
# ¿Qué opináis?

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

**Truqui**: Carga X que debe estar actualmente actualizada a cómo la usaste para entrenar a `model` y haz que `to_pred` sea igual. ¡Sin quitar filas!

Porque:
    - SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 3 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [ ]:
# respuestas

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

**¿De dónde saco `diamonds_test.csv`?**

In [33]:
sample = pd.read_csv("sample_submission.csv")

In [34]:
sample.head()

id  price
0   0  12132
1   1  11786
2   2  14684
3   3  15425
4   4   6724

In [35]:
sample.shape

(13449, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [36]:
submission = pd.DataFrame({"id": range(len(predictions_submit)), "price": predictions_submit})

In [37]:
submission.head()

id        price
0   0   419.550000
1   1  1777.000000
2   2  9852.210000
3   3   510.227238
4   4  8507.760000

In [38]:
submission.shape

(13449, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [39]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.id.all() == sample.id.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [45]:
help(chequeator)

Help on function chequeator in module __main__:

chequeator(df_to_submit)
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.



In [40]:
chequeator(submission)

You're ready to submit!
